<a href="https://colab.research.google.com/github/SojeongShin/NLP_2023/blob/main/13_2_Transformer__BERT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [자연어처리]
# 13주차(13-2). Transformer & BERT

# 202001581 신소정 <- 본인의 학번, 이름으로 변경하여 제출
* **모든 셀 실행** 후 제출하시기 바랍니다.
* **실습 (13-2)**이 있습니다. (제출 기한: 11/28(화) 23시 59분까지)

# 이제 Colab 사용에 모든 수강생분들이 어느정도 익숙해 졌을 거라 생각됩니다!
> ##### 중간 중간 **'### ... 이 부분을 완성하시오'** 라는 부분의 코드를 완성해야 합니다.




# 파이프라인을 이용한 감성분석

In [2]:
# 필요한 라이브러리를 설치하기
### ... 이 부분을 완성하시오
! pip install transformers

In [3]:
# 파이프라인 라이브러리 불러오기
### ... 이 부분을 완성하시오
from transformers import pipeline

# 파이프라인 라이브러리 호출하기: 감성분석
### ... 이 부분을 완성하시오
sentiment_clf = pipeline('sentiment-analysis')

# "what a beautiful day!"라는 문장의 감성 확인하기
### ... 이 부분을 완성하시오
result = sentiment_clf("what a beautiful day!")

print(result)
print("감성분석 결과:", result[0]['label'])
print("감성 스코어:", result[0]['score'])

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

[{'label': 'POSITIVE', 'score': 0.9998812675476074}]
감성분석 결과: POSITIVE
감성 스코어: 0.9998812675476074


In [5]:
from transformers import pipeline
# 오토 토크나이저, 오토 모델(시퀀스 분류) 불러오기
from transformers import AutoTokenizer, AutoModelForSequenceClassification### ... 이 부분을 완성하시오

tokenizer = AutoTokenizer.from_pretrained("monologg/koelectra-base-finetuned-sentiment") ### ... 이 부분을 완성하시오
model = AutoModelForSequenceClassification.from_pretrained("monologg/koelectra-base-finetuned-sentiment") ### ... 이 부분을 완성하시오

sentiment_clf = pipeline("sentiment-analysis", tokenizer=tokenizer, model=model) ### ... 이 부분을 완성하시오

result = sentiment_clf("날이 참 좋아요!")

print(result)
print("감성분석 결과:", result[0]['label'])
print("감성 스코어:", result[0]['score'])

tokenizer_config.json:   0%|          | 0.00/51.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.33k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/279k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/441M [00:00<?, ?B/s]

Some weights of the model checkpoint at monologg/koelectra-base-finetuned-sentiment were not used when initializing ElectraForSequenceClassification: ['classifier.bias', 'classifier.weight']
- This IS expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of ElectraForSequenceClassification were not initialized from the model checkpoint at monologg/koelectra-base-finetuned-sentiment and are newly initialized: ['classifier.dense.weight', 'classifier.out_proj.weight', 'classifier.out_proj.bias', 'classifier.dense.bias']
You should probably TRAIN thi

[{'label': 'positive', 'score': 0.5195437073707581}]
감성분석 결과: positive
감성 스코어: 0.5195437073707581


# 파이프라인을 이용한 질의 응답

In [6]:
context = '''Text mining, also referred to as text data mining (abbr.: TDM),
        similar to text analytics,
        is the process of deriving high-quality information from text. It involves
        "the discovery by computer of new, previously unknown information,
        by automatically extracting information from different written resources."
        Written resources may include websites, books, emails, reviews, and articles.
        High-quality information is typically obtained by devising patterns and trends
        by means such as statistical pattern learning. According to Hotho et al.(2005)
        we can distinguish between three different perspectives of text mining:
        information extraction, data mining, and a KDD (Knowledge Discovery in Databas
        es) process.'''

question = "What is text mining?"

In [8]:
from transformers import pipeline

# 질의응답 파이프라인 호출하기
question_answerer = pipeline("question-answering") ### ... 이 부분을 완성하시오

# 질문과 문단 전달하기
answer = question_answerer(question=question, context=context) ### ... 이 부분을 완성하시오

print(answer)

No model was supplied, defaulted to distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

{'score': 0.42474302649497986, 'start': 110, 'end': 168, 'answer': 'the process of deriving high-quality information from text'}


In [9]:
# 확인하기
context[110:168] ### ... 이 부분을 완성하시오

'the process of deriving high-quality information from text'

In [10]:
context = """수원 화성은 조선시대 화성유수부 시가지를 둘러싼 성곽이다.
1789년(정조 13) 수원을 팔달산 동쪽 아래로 옮기고,
1794년(정조 18) 축성을 시작해 1796년에 완성했다."""
context = context.strip().replace("\n","")

question = "수원 화성은 언제 완성되었는가?"

In [12]:
from transformers import pipeline
# 오토 토크나이저, 오토 모델(질의응답) 불러오기
from transformers import AutoTokenizer, AutoModelForQuestionAnswering ### ... 이 부분을 완성하시오

tokenizer = AutoTokenizer.from_pretrained("monologg/koelectra-small-v2-distilled-korquad-384") ### ... 이 부분을 완성하시오
model = AutoModelForQuestionAnswering.from_pretrained("monologg/koelectra-small-v2-distilled-korquad-384") ### ... 이 부분을 완성하시오

question_answerer = pipeline("question-answering", tokenizer=tokenizer, model=model) ### ... 이 부분을 완성하시오
answer = question_answerer({"question": question, "context": context}) ### ... 이 부분을 완성하시오

print(answer)

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/472 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/255k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/54.8M [00:00<?, ?B/s]

{'score': 0.9962994456291199, 'start': 85, 'end': 91, 'answer': '1796년에'}


# 위키피디아를 이용한 질의응답

In [13]:
# 위키피디아 설치
! pip install wikipedia ### ... 이 부분을 완성하시오

  Preparing metadata (setup.py) ... done
  Created wheel for wikipedia: filename=wikipedia-1.4.0-py3-none-any.whl size=11679 sha256=a94f652b89fd9de8c538a6e060f5eab506c17b808ec65a73a179c90863e1e680
  Stored in directory: /root/.cache/pip/wheels/5e/b6/c5/93f3dec388ae76edc830cb42901bb0232504dfc0df02fc50de
Successfully built wikipedia


In [14]:
import wikipedia
wikipedia.set_lang("ko")

In [15]:
# 제목을 검색(search)하시오
wiki_title = wikipedia.search('뉴진스') ### ... 이 부분을 완성하시오
wiki_title

['최현욱',
 '더 팩트 뮤직 어워즈',
 'NewJeans',
 '박지후',
 '케플러 (음악 그룹)',
 'LG PC 그램',
 '김민지',
 '위버스 (플랫폼)',
 '오스트레일리아',
 '2005년']

In [16]:
# 위키피디아 페이지(page)를 가져오시오.
wiki_text = wikipedia.page('NewJeans').content ### ... 이 부분을 완성하시오
wiki_text

'NewJeans(뉴진스)는 2022년 7월 22일에 데뷔한 대한민국의 5인조 걸 그룹으로, 소속사는 하이브 산하의 레이블인 어도어이다. SM 엔터테인먼트 비주얼 디렉터 출신으로 하이브에 영입된 민희진이 프로듀서로 나서서 발굴한 걸 그룹이다. 2022년 8월 18일에 《엠카운트다운》에서 데뷔 3주만에 첫 1위에 올랐다. 2023년 8월에 잼버리 폐회식 K-POP 공연에 출연했다.\n\n\n== 그룹명 ==\n그룹으로 사용되고 있는 \'뉴진스(NewJeans)\'는 두 가지 의미를 함축하고 있다는 뜻에서 비롯된 말이다. 진스(jeans), 즉 청바지는 시대를 초월한 패션 아이템이라는 생각과 시대를 초월한 이미지를 스스로 개척하려는 그룹의 의도를 암시한다. 이 이름은 또한 새로운 세대(new generation)의 팝 음악을 선도하는 그룹을 가리키는 "new genes"라는 문구의 말장난이기도 하다.\n\n\n== 역사 ==\n\n\n=== 2011~2022: 데뷔 전 활동 및 결성 ===\nNewJeans로 데뷔하기 전에 여러 그룹 멤버가 TV, 음악, 댄스에 참여했다. 호주 태생의 다니엘 마쉬(Danielle Marsh)는 한국에 살 때 2011년까지 방영된 버라이어티 쇼인 tvN 레인보우 유치원의 정규 출연자였다. 그녀는 또한 TV 프로그램인 제시의 플레이 키친과 내 마음의 크레파스에도 출연했다. 혜인은 2017년 11월 동요그룹 우쏘걸(U.SSO Girl)의 멤버 유정(U.Jeong)으로 데뷔했다가 1년 만에 팀에서 탈퇴했다. 2020년 12월, 그녀는 PocketTV가 결성한 음악 그룹이자 유튜브 그룹인 Play With Me Club에 가입했고, 2021년 5월 3일에 그룹을 졸업했다. 베트남계 호주인 하니(Hanni Pham)는 Aemina Dance Crew의 멤버로 멜버른에서 공연을 시작했다. 하니와 민지는 방탄소년단의 2021년 뮤직비디오 \'Permission to Dance\'에 게스트로 출연했다.빅히트 엔터테인먼트와 쏘스 뮤직이 합작한 신인

In [17]:
context = wiki_text
context = context.strip().replace("\n","")
question = "뉴진스는 몇 인조 그룹이야?"

In [18]:
from transformers import ElectraTokenizer, ElectraForQuestionAnswering, pipeline

tokenizer = ElectraTokenizer.from_pretrained("monologg/koelectra-small-v2-distilled-korquad-384")
model = ElectraForQuestionAnswering.from_pretrained("monologg/koelectra-small-v2-distilled-korquad-384")

question_answerer = pipeline("question-answering", tokenizer=tokenizer, model=model)
answer = question_answerer({"question": question, "context": context})

print(answer)

{'score': 0.9953400492668152, 'start': 39, 'end': 42, 'answer': '5인조'}


---

# [[실습 13-2]] 위키피디아 문서를 context로 활용한 질의응답
* 본인의 관심사와 관련있는 위키피디아 페이지를 가져와서, 질의를 해보고 그에 대한 응답 확인해보기
    1. 관심있는 키워드에 대해 검색 후 질의응답 해보는 코드 작성
    2. 결과에 대한 해석을 자유롭게 마크다운으로 작성 (결과 예. {'score': 0.9953400492668152, 'start': 39, 'end': 42, 'answer': '5인조'})


In [25]:
import wikipedia
wikipedia.set_lang("ko")

# 제목을 검색(search)하시오
wiki_title = wikipedia.search('크리스마스 뮤지컬') ### ... 이 부분을 완성하시오
wiki_title

['팀 버튼의 크리스마스 악몽',
 '어 베리 머리 크리스마스',
 '앨런 멩컨',
 '메리대구 공방전',
 '미녀와 야수: 마법의 크리스마스',
 '카이 (1981년)',
 '정상윤',
 '징글 쟁글: 저니의 크리스마스',
 '시시 스페이식',
 '애니 (뮤지컬)']

In [26]:
# 위키피디아 페이지(page)를 가져오시오.
wiki_text = wikipedia.page('애니 (뮤지컬)').content ### ... 이 부분을 완성하시오
wiki_text

'《애니》(Annie)는 1977년 4월 21일, 작곡 찰리 스트라우트가 만든 책 \'어린 고아 애니 (Little Orphan Annie)\'를 원작으로 만든 브로드웨이 뮤지컬이다. 이 뮤지컬에 대표곡 \'tomorrow\'(투모로우)는 많은 사람들의 마음을 울렸다.\n\n\n== 개요 ==\n1976년 미국 코네티컷의 굿스피드 오페라 하우스에서 처음으로 초연한뒤, 이듬해인 1977년 4월 21일, 브로드웨이에서 초연된 애니는 이후 1980년대부터 미국 순회 공연을 하고, 영국 웨스트엔드에 초연되었으며, 호주와 유럽, 일본 등 여러나라에서도 초연되었다. 1977년 토니 어워드 안무상, 대본상, 배우상등 여러개의 상을 휩쓸었고 드라마 데스크 어워드상까지 휩쓸고, 애니 주인공이 여우주연상까지 받으면서 애니의 인기는 날로 커져갔다. 그리고 1997년 브로드웨이 리바이벌 공연 및 1998년 영국 웨스트엔드 리바이벌에 이어 2000년대부터 영국 및 미국 순회공연을 하면서 애니의 인기는 지금까지 이어져오고 있다. 또한 일본의 경우에는 일본의 한 방송국의 주최하에, 1986년부터 지금까지 매해 공연에 오르면서 일본인들의 인기 또한 자랑할 정도이다.\n특히 1막에서 애니가 개에게 부르는 \'Tomorrow\'는 엄청난 인기를 자랑했으며, 그 다음으로 애니와 고아가 부르는 \'Maybe\', 아동노동으로 힘든 생활을 부르는 \'It\'s a hard knock life\' 또한 인기다.\n\n\n== 스토리 ==\n1933년 뉴욕 시립 고아원. 애니에게는 11년 후 찾아오겠다는 부모의 편지와 목걸이만 있을 뿐이다. 애니는 언젠가는 부모님이 찾아오실거라면서 자신의 고아가 아니라고 한다. 그런데 고아원 원장 술 주정뱅이 해니건 원장은 혹독하게 일만 시킨다. 애니는 고아원을 탈출하지만 다시 잡혀 20년 외출 금지라는 벌을 받는다.\n그리고 크리스마스 시즌. 워벅스 씨의 비서 그레이스가 크리스마스 연휴를 함께 보낼 고아를 고르러 왔다가 애니를 데리고 간다. 무뚝뚝한 워벅스씨는 애니

In [27]:
context = wiki_text
context = context.strip().replace("\n","")
question = "애니의 출연진은 누구야?"

In [28]:
from transformers import ElectraTokenizer, ElectraForQuestionAnswering, pipeline

tokenizer = ElectraTokenizer.from_pretrained("monologg/koelectra-small-v2-distilled-korquad-384")
model = ElectraForQuestionAnswering.from_pretrained("monologg/koelectra-small-v2-distilled-korquad-384")

question_answerer = pipeline("question-answering", tokenizer=tokenizer, model=model)
answer = question_answerer({"question": question, "context": context})

print(answer)

{'score': 0.9893172979354858, 'start': 5646, 'end': 5652, 'answer': '아일린 퀸은'}


결과 : {'score': 0.9893172979354858, 'start': 5646, 'end': 5652, 'answer': '아일린 퀸은'}


---